pkg/protocol/gossip/service.go

```diff
@@ -0,0 +1,447 @@
+		inboundStreamChan:   make(chan network.Stream),
+		connectedChan:       make(chan *connectionmsg),
+		disconnectedChan:    make(chan *connectionmsg),
+		streamClosedChan:    make(chan *streamclosedmsg),
+		relationUpdatedChan: make(chan *relationupdatedmsg),
+		streamReqChan:       make(chan *streamreqmsg),
+		forEachChan:         make(chan *foreachmsg),
+	}
+	if s.opts.Logger != nil {
+		s.registerLoggerOnEvents()
+	}
+	return s
+}
+
+// Service handles ongoing gossip streams.
+type Service struct {
+	// Events happening around a Service.
+	Events   ServiceEvents
+	host     host.Host
+	protocol protocol.ID
+	streams  map[peer.ID]*Protocol
+	manager  *p2p.Manager
+	opts     *ServiceOptions
+	// event loop channels
+	inboundStreamChan   chan network.Stream
+	connectedChan       chan *connectionmsg
+	disconnectedChan    chan *connectionmsg
+	streamClosedChan    chan *streamclosedmsg
+	relationUpdatedChan chan *relationupdatedmsg
+	streamReqChan       chan *streamreqmsg
+	forEachChan         chan *foreachmsg
+}
+
+// Protocol returns the gossip.Protocol instance for the given peer or nil.
+func (s *Service) Protocol(id peer.ID) *Protocol {
+	back := make(chan *Protocol)
+	s.streamReqChan <- &streamreqmsg{p: id, back: back}
+	return <-back
+}
+
+// ProtocolForEachFunc is used in Service.ForEach.
+// Returning false indicates to stop looping.
+// This function must not call any methods on Service.
+type ProtocolForEachFunc func(proto *Protocol) bool
+
+// ForEach calls the given ProtocolForEachFunc on each Protocol.
+func (s *Service) ForEach(f ProtocolForEachFunc) {
+	back := make(chan struct{})
+	s.forEachChan <- &foreachmsg{f: f, back: back}
+	<-back
+}
+
+// SynchronizedCount returns the count of streams with peers
+// which appear to be synchronized given their latest Heartbeat message.
+func (s *Service) SynchronizedCount(latestMilestoneIndex milestone.Index) int {
+	var count int
+	s.ForEach(func(proto *Protocol) bool {
+		if proto.IsSynced(latestMilestoneIndex) {
+			count++
+		}
+		return true
+	})
+	return count
+}
+
+// Start starts the Service's event loop.
+func (s *Service) Start(shutdownSignal <-chan struct{}) {
+	s.host.SetStreamHandler(s.protocol, func(stream network.Stream) {
+		s.inboundStreamChan <- stream
+	})
+	s.manager.Events.Connected.Attach(events.NewClosure(func(p *p2p.Peer, conn network.Conn) {
+		s.connectedChan <- &connectionmsg{p: p, conn: conn}
+	}))
+	s.manager.Events.Disconnected.Attach(events.NewClosure(func(p *p2p.Peer) {
+		s.disconnectedChan <- &connectionmsg{p: p, conn: nil}
+	}))
+	s.manager.Events.RelationUpdated.Attach(events.NewClosure(func(p *p2p.Peer, oldRel p2p.PeerRelation) {
+		s.relationUpdatedChan <- &relationupdatedmsg{peer: p, oldRelation: oldRel}
+	}))
+	// manage libp2p network events
+	s.host.Network().Notify((*netNotifiee)(s))
+	s.eventLoop(shutdownSignal)
+	// de-register libp2p network events
+	s.host.Network().StopNotify((*netNotifiee)(s))
+}
+
+type connectionmsg struct {
+	p    *p2p.Peer
```

Comment:
should be consistent naming between these structs. Sometimes its p, sometimes peer.

`p2p.Peer` should be `peer`
and
`peer.ID` should be `peerID`